cmd/snap-update-ns/snap-update-ns.c

```diff
@@ -19,7 +19,299 @@
+#include <sys/mount.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/vfs.h>
+#include <unistd.h>
+
+#include "../libsnap-confine-private/utils.h"
+#include "../libsnap-confine-private/error.h"
+#include "../libsnap-confine-private/cleanup-funcs.h"
+#include "../libsnap-confine-private/mountinfo.h"
+#include "mount-entry.h"
+
+static void missing_locking()
+{
+	fprintf(stderr, "XXX: snap-alter-ns doesn't lock"
+		" the mount namespace yet.\n");
+}
+
+static void sc_reassociate_with_snap_namespace_or_exit(const char *snap_name);
+static bool sc_should_act_on_change(const struct sc_mount_change *change);
+
+#define SC_DESIRED_PROFILE_FMT "/var/lib/snapd/mount/snap.%s.fstab"
+#define SC_CURRENT_PROFILE_FMT "/run/snapd/ns/%s.fstab"
+#define SC_MNT_NS_FMT          "/run/snapd/ns/%s.mnt"
+
+int main(int argc, char **argv)
+{
+	if (argc != 2) {
+		printf("Usage: snap-alter-ns SNAP-NAME");
+		return 1;
+	}
+	const char *snap_name = argv[1];
+	// TODO: verify once verify_snap_name lands.
+
+	debug("Checking if the mount namespace of snap %s needs changes",
+	      snap_name);
+
+	// TODO: use locking from ns-support.
+	//
+	// This ensures we see consistent "current" and "desired" profiles.
+	//
+	// The current profile is modified by snap-discard-ns, snap-alter-ns and
+	// snap-confine. All the tools follow the locking system.
+	//
+	// The desired profile is modified by snapd. Snapd runs snap-alter-ns we
+	// put the burden of not clobbering this file while we may be reading.
+	missing_locking();
+
+	// The desired profile is stored in /var/lib/snapd/mount/$SNAP_NAME.fstab
+	// The current profile is stored in /run/snapd/ns/$SNAP_NAME.fstab
+	//
+	// We are loading both to compare them and compute what needs to be done to
+	// alter the namespace to match the desired profile.
+	char buf[PATH_MAX];
+	struct sc_mount_entry
+	    __attribute__ ((cleanup(sc_cleanup_mount_entry_list))) * desired =
+	    NULL;
+	must_snprintf(buf, sizeof buf, SC_DESIRED_PROFILE_FMT, snap_name);
+	desired = sc_load_mount_profile(buf);
+	debug("Loaded desired mount profile:");
+	for (struct sc_mount_entry * e = desired; e != NULL; e = e->next) {
+		debug("\tdesired: %s", sc_mount_entry_to_string(e));
+	}
+	if (desired == NULL) {
+		debug("\tdesired: (empty profile)");
+	}
+
+	struct sc_mount_entry
+	    __attribute__ ((cleanup(sc_cleanup_mount_entry_list))) * current =
+	    NULL;
+	must_snprintf(buf, sizeof buf, SC_CURRENT_PROFILE_FMT, snap_name);
+	current = sc_load_mount_profile(buf);
+	debug("Loaded current mount profile");
+	for (struct sc_mount_entry * e = current; e != NULL; e = e->next) {
+		debug("\tcurrent: %s", sc_mount_entry_to_string(e));
+	}
+	if (current == NULL) {
+		debug("\tcurrent: (empty profile)");
+	}
+
+	if (current == NULL && desired == NULL) {
+		debug("There's nothing to do");
+		return 0;
+	}
+	// TODO: correct the wiki, we don't quit if something is not present as
+	// this is a valid case as well (e.g. a profile gets removed).
```

Comment:
The use cases are:
- neither exists: nothing to do
- current exists but desired does not: unmount everything and remove current (profile removal)
- current does not exist, but desired does: mount everything and create current (first run after profile add)
- both exist: mount/umount as necessary and update current (subsequent updates not including profile removal)